addons/stock/models/stock_rule.py

```diff
@@ -558,6 +558,27 @@ def _search_rule(self, route_ids, packaging_uom_id, product_id, warehouse_id, do
                 res = Rule.search(expression.AND([[('route_id', 'in', warehouse_routes.ids)], domain]), order='route_sequence, sequence', limit=1)
         return res
 
+    def _get_product_routes(self, product):
+        """ Get valid routes for a product based on the following conditions:
+            - If `purchase_ok` is enabled and the product has a vendor, select the 'Buy' route.
+            - If the product has a Bill of Materials (BOM), select the 'Manufacturing' route.
+            - If the product is a component of a subcontracting BOM and has sufficient stock,
+            select the 'Resupply Subcontractor on Order' route.
+        """
+        if not product:
+            return self.env["stock.route"]
+
+        buy_route = self.env.ref("purchase_stock.route_warehouse0_buy", raise_if_not_found=False)
+        manufacturing_route = self.env.ref("mrp.route_warehouse0_manufacture", raise_if_not_found=False)
+        route_ids = product.categ_id.total_route_ids | product.route_ids
+
+        if buy_route and product.purchase_ok and product.seller_ids:
```

Comment:
buy_route: condition isn't strictly necessary, and if it is None, None won't be added to route_ids, so the check is technically not required.